### PR TITLE
Throw exception when trying to use non-backed enum types

### PR DESCRIPTION
**Pull request by thePanz**
This solves #11173

### CODE REVIEW

**Comment by greg0ire** [comment]
It seems there are several CI jobs failing. Please take a look at [this guide](https://www.doctrine-project.org/contribute/#dealing-with-checks-and-tools) for more on how to handle those.

**Comment by thePanz** [comment]
@greg0ire Pushed some more fixes.

I did not run phpunit still as I do not have all the required DB backends tho

**Comment by greg0ire** [comment]
You only need SQLite to run most tests, and you need nothing at all to run the tests you wrote.

File: lib/Doctrine/ORM/Mapping/DefaultTypedFieldMapper.php [hunk 1]
```diff
@@ -57,7 +57,17 @@ public function validateAndComplete(array $mapping, ReflectionProperty $field):
                 $mapping['enumType'] = $type->getName();
 
                 $reflection = new ReflectionEnum($type->getName());
-                $type       = $reflection->getBackingType();
+                if (
+                    $reflection->isBacked() === false
+                ) {
```
Comment (greg0ire) [inline lib/Doctrine/ORM/Mapping/DefaultTypedFieldMapper.php]:
```suggestion
                if(! $reflection->isBacked()) {
```

File: tests/Doctrine/Tests/ORM/Mapping/TypedEnumFieldMapperTest.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        return new DefaultTypedFieldMapper();
+    }
+
+    public function testNotBackedEnumThrows(): void
+    {
+        $reflectionClass = new ReflectionClass(FaultySwitch::class);
+
+        $this->expectException(MappingException::class);
+        $this->expectExceptionMessage(sprintf(
```
Comment (greg0ire) [inline tests/Doctrine/Tests/ORM/Mapping/TypedEnumFieldMapperTest.php]: I don't think you should use `sprintf` here. Please show the final exception message, as the user will see it instead.

File: tests/Doctrine/Tests/Models/Enums/FaultySwitch.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+{
+    #[Column(type: 'string')]
+    public string $value;
+
+    /**
+     * The following line is ignored on psalm and phpstan so that we can test
+     * that the mapping is throwing an exception when a non-backed enum is used.
+     *
+     * @psalm-suppress InvalidArgument
+     */
```
Comment (greg0ire) [inline tests/Doctrine/Tests/Models/Enums/FaultySwitch.php:19]: Ah so this means we already have static analysis for this kind of issue. Good. I suppose it cannot hurt to also validate it at runtime.

Comment (thePanz) [reply tests/Doctrine/Tests/Models/Enums/FaultySwitch.php:19]: yes, the library is covered by phpstan and psalm
Better having a runtime check as 3rd party users might not have the static analysis in place, or the static-analysis did not run yet (like my case) :)

**Review by greg0ire — APPROVED** [review]

**Review by SenseException — APPROVED** [review]

**Pull request merged**